_posts/2021-06-22-type-phantoms-and-safe-apis.md

```diff
@@ -0,0 +1,712 @@
+use `api.getShips(undefined, undefined)` to fetch ships. This can be remedied with a more complex,
+conditional `Handler` type, and a type cast:
+
+```typescript
+// lib/Client.ts
+
+type Handler<EP extends Endpoint> =
+  EP['body'] extends object
+    ? HandlerFull<EP>
+    : EP['path'] extends object
+    ? HandlerPath<EP>
+    : HandlerNone<EP>
+type HandlerFull<EP extends Endpoint> =
+  (params: EP['path'], body: EP['body']) => Promise<EP['result']>
+type HandlerPath<EP extends Endpoint> = (params: EP['path']) => Promise<EP['result']>
+type HandlerNone<EP extends Endpoint> = () => Promise<EP['result']>
+```
+
+Now it's possible to call `api.getShips()`, but an extra `undefined` is still required when there is
+a body parameter, but no path parameters: `api.addShip(undefined, features)`.
+
+### Adding query parameters
+
+Adding query parameters would be straightforward: just add a `query` property to the common `Call`
+type. Quite a few of the utilities would need changes though, and the new property would need to be
+added to all endpoint declarations.
+
+With query parameters added, using a single parameter object for the client might be more ergonomic:
+
+```typescript
+api.getShips()
+api.addShip({ body: features })
+api.searchShips({ query: { name: 'tie fighter' } })
+```
+
+### Implementing error handling
+
+The basic example doesn't deal with _any_ errors. In a real application both the server and the
+client need some error handling.
+
+If a server handler throws an exception on failure, we can use the exception to produce an error
+response:
+
+```typescript
+// lib/Server.ts
+
+const createHandler = <EP extends Endpoint>(handler: Handler<EP>): RouteHandler<EP> =>
+  async (req, res) => {
+    try {
+      const result = await handler({ path: req.params, body: req.body })
+
+      res.send(result)
+    } catch (error) {
+      res.sendStatus(500)
+    }
+  }
+```
+
+This can be expanded to check error types and produce different status codes based on that. An error
+object library like [boom](https://hapi.dev/module/boom/) could be useful as well.
+
+An interesting alternative is to use more functional style, and a `Result` / `Either` type.
+Languages like [F#](https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/results),
+[Haskell](https://hackage.haskell.org/package/base/docs/Data-Either.html)
+and [Rust](https://doc.rust-lang.org/std/result/enum.Result.html) use them extensively.
+
+Result types
+are [extremely useful in TypeScript](https://imhoff.blog/posts/using-results-in-typescript)
+as well, and a bare-bones implementation is easy:
+
+```typescript
+export type Result<Success, Failure> =
+  { ok: true; value: Success } | { ok: false; error: Failure }
+```
+
+One way to use them on the server-side would be to expand the `Handler` type:
+
+```typescript
+// lib/Server.ts
+
+export type Handler<EP extends Endpoint> =
+  (params: Params<EP>) => Promise<Result<EP['result'], string>>
```

Comment:
Cool to see the mention of the `Result` type! This probably isn't the place to dive _too_ deep, but wouldn't this be a good place to showcase / mention typed errors as well? IMO much of the benefit of using something like `Result` or `Either` is that you get to type the error channel as well, instead of it implicitly being any throwable (or here any `string` error message). For example, our neighbours at Supermetrics had [this pretty interesting take on it](https://dev.to/supermetrics/simple-and-maintainable-error-handling-in-typescript-56lm).

This is definitely out-of-scope, but there's also an approach where failure modes are split into expected domain errors (via a typed error channel) and unexpected defects (that are thrown without local handling). At least the ZIO (and Effect-TS) ecosystem does that.

Reply:
I could add a note about expanding the error channel type from a simple `string`. We've actually used that as well.

There's also a short note in the validation section, that includes an error channel type that's dependent on the success type.

Reply:
Added a note about expanding the failure type.